sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanBuckets.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.extensions.zpages;
+
+import com.google.common.collect.EvictingQueue;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.Status.CanonicalCode;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.Map;
+
+final class TracezSpanBuckets {
+  private static final int NUM_SAMPLES_PER_LATENCY_BUCKET = 10;
+  private static final int NUM_SAMPLES_PER_ERROR_BUCKET = 5;
+
+  private final Map<LatencyBoundaries, EvictingQueue<ReadableSpan>> latencyBuckets =
+      new HashMap<>();
+  private final Map<CanonicalCode, EvictingQueue<ReadableSpan>> errorBuckets = new HashMap<>();
+
+  public TracezSpanBuckets() {
+    for (LatencyBoundaries bucket : LatencyBoundaries.values()) {
+      latencyBuckets.put(
+          bucket, EvictingQueue.<ReadableSpan>create(NUM_SAMPLES_PER_LATENCY_BUCKET));
+    }
+    for (CanonicalCode code : CanonicalCode.values()) {
+      if (!code.toStatus().isOk()) {
+        errorBuckets.put(code, EvictingQueue.<ReadableSpan>create(NUM_SAMPLES_PER_ERROR_BUCKET));
+      }
+    }
+  }
+
+  public void addToBucket(ReadableSpan span) {
+    Status status = span.toSpanData().getStatus();
+    if (status.isOk()) {
+      long latency = span.getLatencyNanos();
+      for (LatencyBoundaries bucket : LatencyBoundaries.values()) {
+        if (latency >= bucket.getLatencyLowerBound() && latency < bucket.getLatencyUpperBound()) {
+          latencyBuckets.get(bucket).add(span);
+          return;
+        }
+      }
+    }
+    errorBuckets.get(status.getCanonicalCode()).add(span);
+  }
+
+  public Map<LatencyBoundaries, Integer> getLatencyBoundariesToCountMap() {
```

Comment:
If we go with this, should I remove `getSpanLatencyCounts(long lowerBound, long upperBound)`? I initially created this function as I wasn't sure how much filtering the package needed.

Reply:
So while implementing this, I've run into the issue of separating OK spans from error spans. The `TracezSpanBuckets` class was implemented to make this simpler, so if we want to incrementally maintain the count maps, then it makes more sense to have the `TracezSpanBuckets` class keep track of the counts, as it's already tracking the spans. I still am not sure whether this is entirely necessary, as creating the count map from the spans is not difficult (see `getLatencyBoundariesToCountMap()` in `TracezSpanBuckets`). I'd appreciate other opinions on this!

Reply:
Since `getSpanLatencyCounts(long lowerBound, long upperBound)` seems unused seems fine to remove it.

> TracezSpanBuckets class keep track of the counts, as it's already tracking the spans

The idea is for `TracezSpanBuckets` to not track spans - if we do that, we get complete metrics (instead of being bounded to 10 or whatever constant we end up with) and it means spans can get cleaned up as quickly as possible, we'd only be keeping track of counts here not whole spans which have lots of data associated with them and even with the capped map is waste of resources IMO.

Though before doing that, I'd suggest also investigating the OpenTelemetry metrics API to see if it has anything we can use. It might not, but I'm suspecting we haven't taken a good look yet. For example, the buckets we're implementing seem suspiciously similar to a histogram aggregation

https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/metrics/view/Aggregations.java#L66

I think it should be possible to record latencies as doubles with the span name as a label and use aggregations to power our views. If OTel can't dogfood it's own metrics API we're in some trouble 😅

Reply:
Histogram aggregation doesn't seem to be implemented for OpenTelemetry yet (the class is calling no-op functions). If I'm understanding your suggestion correctly, you want to replace the buckets class with data structures in the span processor class that individually track the metrics. This would reduce the size of the data stored, as spans are large objects, and allow for more than just the most recent 10/5 spans.

The problem I have with this is that the /tracez zPage needs pretty much all of the span information. The zPage supports a details section that displays the ID's, events, etc. of the spans within a cell of the table. I don't think it makes much sense to track all of these individually instead of the span itself, as I can't imagine much memory would be saved. But perhaps I'm missing something here. Could I schedule a call to go over your thoughts?

Reply:
Yeah, sorry about that. I've been meaning to update my design doc. Will definitely do so before adding it to the README.